interfaces/builtin/shared_memory.go

```diff
@@ -198,37 +228,93 @@ func writeSharedMemoryPaths(w io.Writer, slot *interfaces.ConnectedSlot,
 }
 
 func (iface *sharedMemoryInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
+	privateAttr, isSet := plug.Attrs["private"]
+	private, ok := privateAttr.(bool)
+	if isSet && !ok {
+		return fmt.Errorf(`shared-memory "private" attribute must be a bool, not %v`, privateAttr)
+	}
+	if plug.Attrs == nil {
+		plug.Attrs = make(map[string]interface{})
+	}
+	plug.Attrs["private"] = private
+
 	sharedMemoryAttr, isSet := plug.Attrs["shared-memory"]
 	sharedMemory, ok := sharedMemoryAttr.(string)
 	if isSet && !ok {
 		return fmt.Errorf(`shared-memory "shared-memory" attribute must be a string, not %v`,
 			plug.Attrs["shared-memory"])
 	}
-	if sharedMemory == "" {
-		if plug.Attrs == nil {
-			plug.Attrs = make(map[string]interface{})
+	if private {
+		if isSet {
+			return fmt.Errorf(`shared-memory "shared-memory" attribute must not be set together with "private: true"`)
+		}
+		// A private shared-memory plug cannot coexist with
+		// other shared-memory plugs/slots.
+		for _, other := range plug.Snap.Plugs {
```

Comment:
are we sure that `plug.Snap.Plugs` includes the full set of plugs that a snap uses? what about scoped plug declaration for hooks and apps? I have just grown hesitant of ifacestate doing the right thing all the time, so it would be good if we add a test case for those in the unit tests 

to be clear I am referring to plug declaration like this:
```yaml
name: foo
plugs: # global
  foo:
    interface: shared-memory
    private: true
hooks:
  configure:
    plugs:
      - shared-memory # scoped
```

I'm not sure how you could get a `shared-memory` plug without the `shared-memory` attribute set for it to something other than "" without declaring it globally, but it is tricky to reason about this with great certainty

Reply:
In the `infoFromSnapYaml` function, it calls `setAppsFromSnapYaml` and `setHooksFromSnapYaml` before `SanitizePlugsSlots`:

https://github.com/snapcore/snapd/blob/b70a9f446c9490f9d0548313bab34cbcf167419d/snap/info_snap_yaml.go#L187-L191
https://github.com/snapcore/snapd/blob/b70a9f446c9490f9d0548313bab34cbcf167419d/snap/info_snap_yaml.go#L229

Those first two functions create missing plug and slot definitions referenced by the apps and hooks respectively, so we should see all plugs/slots from the interface's `BeforeSanitizePlug` method.